File: core/src/com/mygdx/game/handler/HandlerChain.java
package com.mygdx.game.handler;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class HandlerChain {
    private List<BaseHandler> chians = new ArrayList<>();
    private int index = 0;
    private ExecutorService service = Executors.newCachedThreadPool();

    public HandlerChain addHandler(BaseHandler IHandler) {
        chians.add(IHandler);
        return this;
    }

    public void start(BaseHandler.HandlerEntity entity, Runnable run) {
        service.submit(new Runnable() {
            @Override
            public void run() {
                process(entity);
                System.out.println("任务链执行完毕");
                // 触发回调
                run.run();
            }
        });
    }

    public BaseHandler.HandlerEntity process(BaseHandler.HandlerEntity entity) {
        if (index >= chians.size()) {
            return entity;
        } else {
            BaseHandler IHandler = chians.get(index);
            index++;
            ParcelData target = entity.getParcelData();
            if (IHandler.isNeedHandle(target.getTargetHandlerName())) {
                System.out.println("执行" + IHandler.getClass().getSimpleName());
                BaseHandler.HandlerEntity result = IHandler.doHandle(entity, this);
                return result;
            } else {
                return process(entity);
            }
        }
    }


    public void reset() {
        index = 0;
    }

}

File: core/src/main/kotlin/com/rich/diy/game/handler/HandlerChain.kt
package com.rich.diy.game.handler

import com.rich.diy.game.handler.BaseHandler.HandlerEntity
import java.util.*
import java.util.concurrent.Executors

class HandlerChain {
    private val chians: MutableList<BaseHandler> = ArrayList()
    private var index = 0
    private val service = Executors.newCachedThreadPool()
    fun addHandler(IHandler: BaseHandler): HandlerChain {
        chians.add(IHandler)
        return this
    }

    fun start(entity: HandlerEntity, run: Runnable) {
        service.submit {
            process(entity)
            println("任务链执行完毕")
            // 触发回调
            run.run()
        }
    }

    fun process(entity: HandlerEntity): HandlerEntity? {
        return if (index >= chians.size) {
            entity
        } else {
            val IHandler = chians[index]
            index++
            val target = entity.parcelData
            if (IHandler.isNeedHandle(target!!.targetHandlerName)) {
                println("执行" + IHandler.javaClass.simpleName)
                IHandler.doHandle(entity, this)
            } else {
                process(entity)
            }
        }
    }

    fun reset() {
        index = 0
    }
}